File: day18/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = Deno.args.reduce((acc, item) => (item == '-d' ? 1 : item == '-d2' ? 2 : item == '-d3' ? 3 : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) =>
    item == '-i' ? 'input.txt' : item == '-e' ? 'example.txt' : item.startsWith('-e') ? `example${item.slice(-1)}.txt` : acc,
  'example.txt'
);
console.log({ FILENAME, DEBUG });

const text = await Deno.readTextFile(FILENAME);
if (DEBUG > 1) console.debug({ text });

const blocks = text
  .split('\n')
  .filter((line) => line.trim())
  .map((line) => line.split(',').map((item) => parseInt(item)));
if (DEBUG) console.debug({ blocks });

const genMap = (lenX: number, lenY: number, blocks: number[][]) => {
  const mapData = Array.from({ length: lenX }, () => Array.from({ length: lenY }).fill('.')) as string[][];
  for (const block of blocks) {
    mapData[block[0]][block[1]] = '#';
  }
  return mapData;
};

const drawMap = (mapData: string[][]) => {
  const lenX = mapData[0].length;
  const lenY = mapData.length;
  for (let y = 0; y < lenY; y++) {
    const row = [`${String(y).padStart(3, ' ')}:  `];
    for (let x = 0; x < lenX; x++) {
      row.push(mapData[x][y]);
    }
    console.debug(row.join(''));
  }
};

class Coord {
  public hash: string;
  constructor(public x: number, public y: number) {
    this.hash = `${x},${y}`;
    // this.hash = JSON.stringify([x, y]);
  }
}

class Path {
  public history: Map<string, Coord> = new Map();
  public current!: Coord;
  constructor(public start: Coord) {
    this.push(start);
  }
  len() {
    return this.history.size - 1;
  }
  push(coord: Coord) {
    if (this.history.has(coord.hash)) return false;
    this.history.set(coord.hash, coord);
    this.current = coord;
    return true;
  }
  clone() {
    const newPath = new Path(this.start);
    newPath.history = new Map(this.history.entries());
    // const histItems = Array.from(this.history.entries());
    // const newPath = new Path(histItems[0][1]);
    // for (const histItem of histItems.slice(1)) {
    //   newPath.push(histItem[1]);
    // }
    return newPath;
  }
}

const getShortestPath = (mapData: string[][], startX: number, startY: number, endX: number, endY: number) => {
  const lenX = mapData[0].length;
  const lenY = mapData.length;
  const end = new Coord(endX, endY);
  let paths = [new Path(new Coord(startX, startY))];
  const shortestPaths: Map<string, number> = new Map();
  const nextPaths: Path[] = [];
  for (let i = 0; i < lenX * lenY; i++) {
    for (const path of paths) {
      for (let dir = 0; dir < 4; dir++) {
        const nextCoord = new Coord(
          path.current.x + (dir == 0 ? -1 : dir == 2 ? 1 : 0),
          path.current.y + (dir == 3 ? -1 : dir == 1 ? 1 : 0)
        );

        // oob or block
        if (
          nextCoord.x < 0 ||
          nextCoord.x >= lenX ||
          nextCoord.y < 0 ||
          nextCoord.y >= lenY ||
          mapData[nextCoord.x][nextCoord.y] == '#'
        )
          continue;

        // get/set shortestPaths, continue if not shortest
        const shortestPath = shortestPaths.get(nextCoord.hash);
        if (typeof shortestPath === 'undefined') {
          shortestPaths.set(nextCoord.hash, path.len());
        } else if (shortestPath <= path.len()) {
          continue;
        }
        // TODO: don't think we need to handle shortestPath>path.len

        const nextPath = path.clone();
        // returns false if duplicate
        if (!nextPath.push(nextCoord)) continue;

        // return the completed path immediately
        if (nextCoord.hash == end.hash) {
          return nextPath;
        } else {
          nextPaths.push(nextPath);
        }
      }
    }
    paths = nextPaths;
  }
  throw new Error('no path found');
};

const part1 = (blocks: number[][]) => {
  let lenX: number, lenY: number, blockCount: number;
  switch (FILENAME) {
    case 'input.txt':
      lenX = 71;
      lenY = 71;
      blockCount = 1024;
      break;
    case 'example.txt':
      lenX = 7;
      lenY = 7;
      blockCount = 12;
      break;
    default:
      throw new Error(`add lenX,lenY,blockCount case for filename ${FILENAME}`);
  }
  const mapData = genMap(lenX, lenY, blocks.slice(0, blockCount));
  if (DEBUG) drawMap(mapData);
  const shortestPath = getShortestPath(mapData, 0, 0, lenX - 1, lenY - 1);
  console.log('part 1', shortestPath.len());
};

const part2 = (blocks: number[][]) => {
  let lenX: number, lenY: number;
  switch (FILENAME) {
    case 'input.txt':
      lenX = 71;
      lenY = 71;
      break;
    case 'example.txt':
      lenX = 7;
      lenY = 7;
      break;
    default:
      throw new Error(`add lenX,lenY case for filename ${FILENAME}`);
  }
  //TODO: probably quicker to loop backwards?
  // for (let blockCount = 1024; blockCount < blocks.length; blockCount++) {
  for (let blockCount = blocks.length; blockCount > 0; blockCount--) {
    const blockSlice = blocks.slice(0, blockCount);
    console.log({ blockCount, len: blockSlice.length });
    const mapData = genMap(lenX, lenY, blockSlice);
    if (DEBUG) drawMap(mapData);
    try {
      const shortestPath = getShortestPath(mapData, 0, 0, lenX - 1, lenY - 1);
      console.log('part 2', blockCount + 1, blocks.slice(blockCount, blockCount + 1)[0]);
      break;
    } catch {
      // console.log('part 2', blockCount, blocks.slice(blockCount - 1, blockCount)[0]);
      // break;
    }
  }
};

part1(blocks);
part2(blocks);

File: day18/solution.go
//go:build ignore

package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type Coord struct {
	x, y int
}

func main() {
	filename, debug := parseArgs()
	fmt.Printf("filename: %s; debug: %d\n", filename, debug)
	blocks := readData(filename)
	switch debug {
	case 0:
		break
	case 1:
		fmt.Println("len blocks", len(blocks))
	default:
		fmt.Println(blocks)
	}
	lenX := 0
	lenY := 0
	numBlocks := 0
	switch filename {
	case "example.txt":
		lenX = 7
		lenY = 7
		// numBlocks = 12
		numBlocks = 3
	case "input.txt":
		lenX = 71
		lenY = 71
		numBlocks = 1024
	}
	part1(lenX, lenY, blocks[0:numBlocks], debug)
	part2(lenX, lenY, blocks, debug)
}

func part2(lenX int, lenY int, blocks []Coord, debug int) {
	startAt := Coord{0, 0}
	endAt := Coord{lenX - 1, lenY - 1}
	for i := len(blocks); i > 0; i-- {
		mapData := genMap(lenX, lenY, blocks[0:i])
		if debug > 0 {
			drawMap(mapData)
		}
		pathLength := getShortestPath(mapData, startAt, endAt, debug)
		if pathLength > -1 {
			blockingCoord := blocks[i]
			fmt.Println("part 2:", i+1, blockingCoord)
			break
		}
	}
}

func part1(lenX int, lenY int, blocks []Coord, debug int) {
	mapData := genMap(lenX, lenY, blocks)
	if debug > 0 {
		drawMap(mapData)
	}
	startAt := Coord{0, 0}
	endAt := Coord{lenX - 1, lenY - 1}
	pathLength := getShortestPath(mapData, startAt, endAt, debug)
	fmt.Println("part 1:", pathLength)
}

func getShortestPath(mapData [][]bool, startAt Coord, endAt Coord, debug int) int {
	lenX := len(mapData[0])
	lenY := len(mapData)
	paths := [][]Coord{{startAt}}
	coordDistances := map[Coord]int{}
	for i := 0; i < lenX*lenY; i++ {
		if debug > 0 {
			fmt.Printf("i=%d len=%d\n", i, len(paths))
		}
		newPaths := [][]Coord{}
		for _, path := range paths {
			for d := 0; d < 4; d++ {
				prevCoord := path[i]
				coord := Coord{prevCoord.x, prevCoord.y}
				switch d {
				case 0:
					coord.y--
				case 1:
					coord.x++
				case 2:
					coord.y++
				case 3:
					coord.x--
				}
				// continue on oob, block, in path
				if coord.x < 0 || coord.x >= lenX || coord.y < 0 || coord.y >= lenY || mapData[coord.x][coord.y] || inPath(path, coord) {
					continue
				}
				coordDistance := coordDistances[coord]
				pathLen := i + 1
				if coordDistance == 0 {
					// set coord distance if none exists
					coordDistances[coord] = pathLen
				} else if coordDistance <= pathLen {
					// continue if our path is not shorter
					continue
				}
				if coord == endAt {
					// first to reach the end
					return pathLen
				}
				// otherwise, push the new path onto new paths
				newPath := make([]Coord, i+2)
				copy(newPath, path)
				newPath[pathLen] = coord
				newPaths = append(newPaths, newPath)
			}
		}
		paths = newPaths
	}
	return -1
}

func inPath(path []Coord, search Coord) bool {
	for _, i := range path {
		if i == search {
			return true
		}
	}
	return false
}

func drawMap(mapData [][]bool) {
	lenX := len(mapData[0])
	lenY := len(mapData)
	row := make([]string, lenX+1)
	for y := 0; y < lenY; y++ {
		row[0] = fmt.Sprintf("% 3d:  ", y)
		for x := 0; x < lenX; x++ {
			if mapData[x][y] {
				row[x+1] = "#"
			} else {
				row[x+1] = "."
			}
		}
		fmt.Println(strings.Join(row, ""))
	}
}

func genMap(lenX int, lenY int, blocks []Coord) [][]bool {
	data := make([][]bool, lenX)
	for i := 0; i < lenX; i++ {
		data[i] = make([]bool, lenY)
	}
	for _, block := range blocks {
		data[block.x][block.y] = true
	}
	return data
}

func readData(filename string) []Coord {
	content, _ := os.ReadFile(filename)
	regex, _ := regexp.Compile("(-?[0-9]+)")
	matches := regex.FindAllStringSubmatch(string(content), -1)
	data := []Coord{}
	fields := make([]int, 2)
	for i, match := range matches {
		fields[i%len(fields)], _ = strconv.Atoi(match[0])
		if i%len(fields) == len(fields)-1 {
			data = append(data, Coord{
				fields[0],
				fields[1],
			})
		}
		// fmt.Println("i",i,"match",match,fields)
	}
	return data
}

func parseArgs() (filename string, debug int) {
	filename = "example.txt"
	debug = 0
	for _, arg := range os.Args[1:] {
		switch arg {
		case "-i":
			filename = "input.txt"
		case "-e2":
			filename = "example2.txt"
		case "-e3":
			filename = "example3.txt"
		case "-e4":
			filename = "example4.txt"
		case "-d":
			debug = 1
		case "-d2":
			debug = 2
		case "-d3":
			debug = 3
		default:
			fmt.Printf("unknown arg: %s\n", arg)
		}
	}
	return
}
